From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Sun, 7 Mar 2021 16:41:05 -0500
Subject: [PATCH] Enhanced Enderman mechanics and config options

- Allows Enderman to target (aggro) Endermites that were spawned using a command or Spawn Egg.
- Treat players wearing the Ender Dragon head as if they're wearing a pumpkin.
- Disable the player stare aggression mechanic.

diff --git a/src/main/java/net/minecraft/server/EntityEnderman.java b/src/main/java/net/minecraft/server/EntityEnderman.java
index b884addf2ce6f1ef7394658078deb2e75370654f..4c325478666517b28830d789ebd58548ab28f161 100644
--- a/src/main/java/net/minecraft/server/EntityEnderman.java
+++ b/src/main/java/net/minecraft/server/EntityEnderman.java
@@ -16,7 +16,7 @@ public class EntityEnderman extends EntityMonster implements IEntityAngerable {
     private static final DataWatcherObject<Boolean> bo = DataWatcher.a(EntityEnderman.class, DataWatcherRegistry.i);
     private static final DataWatcherObject<Boolean> bp = DataWatcher.a(EntityEnderman.class, DataWatcherRegistry.i);
     private static final Predicate<EntityLiving> bq = (entityliving) -> {
-        return entityliving instanceof EntityEndermite && ((EntityEndermite) entityliving).isPlayerSpawned();
+        return entityliving instanceof EntityEndermite && (((EntityEndermite) entityliving).isPlayerSpawned() || entityliving.world.purpurConfig.endermanAggroSpawnedEndermites); // Purpur
     };
     private int br = Integer.MIN_VALUE;
     private int bs;
@@ -197,7 +197,7 @@ public class EntityEnderman extends EntityMonster implements IEntityAngerable {
         // Paper end
         ItemStack itemstack = (ItemStack) entityhuman.inventory.armor.get(3);
 
-        if (itemstack.getItem() == Blocks.CARVED_PUMPKIN.getItem()) {
+        if (world.purpurConfig.endermanDisableStareMechanic || itemstack.getItem() == Blocks.CARVED_PUMPKIN.getItem() || (world.purpurConfig.endermanIgnoreEnderDragonHead && itemstack.getItem() == Items.DRAGON_HEAD)) { // Purpur
             return false;
         } else {
             Vec3D vec3d = entityhuman.f(1.0F).d();
diff --git a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
index bf8a842e7c43946c87c8f27afdf364b1475cf44e..00fb616bf6d71fa3f5a1c4100f329f274dba1de7 100644
--- a/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
+++ b/src/main/java/net/pl3x/purpur/PurpurWorldConfig.java
@@ -811,6 +811,9 @@ public class PurpurWorldConfig {
     public boolean endermanDespawnEvenWithBlock = false;
     public double endermanMaxHealth = 40.0D;
     public boolean endermanTakeDamageFromWater = true;
+    public boolean endermanAggroSpawnedEndermites = false;
+    public boolean endermanIgnoreEnderDragonHead = false;
+    public boolean endermanDisableStareMechanic = false;
     private void endermanSettings() {
         endermanRidable = getBoolean("mobs.enderman.ridable", endermanRidable);
         endermanRidableInWater = getBoolean("mobs.enderman.ridable-in-water", endermanRidableInWater);
@@ -824,6 +827,9 @@ public class PurpurWorldConfig {
         }
         endermanMaxHealth = getDouble("mobs.enderman.attributes.max_health", endermanMaxHealth);
         endermanTakeDamageFromWater = getBoolean("mobs.enderman.takes-damage-from-water", endermanTakeDamageFromWater);
+        endermanAggroSpawnedEndermites = getBoolean("mobs.enderman.aggressive-towards-spawned-endermites", endermanAggroSpawnedEndermites);
+        endermanIgnoreEnderDragonHead = getBoolean("mobs.enderman.ignore-players-wearing-ender_dragon-head", endermanIgnoreEnderDragonHead);
+        endermanDisableStareMechanic = getBoolean("mobs.enderman.disable-player-stare-mechanic", endermanDisableStareMechanic);
     }
 
     public boolean endermiteRidable = false;
